File: JS/Matrix.js
function Matrix(y,x){
	this.dimX = x;//int (columns)
	this.dimY = y;//int (rows)
	this.val = new Array();//array(array(),array(),ect.)
	

	this.isMatrix = function(y,x){//int,int->bool
		if(typeof(this)!="object"){
			return false;
		}else if(this.val.length != y){
			return false;
		}else{
			check = true;
			for(var i of this.val){
				if(i.length!=x){
					check = false;
				}else{continue;}
			}
			return check;
		}
	}

	this.clone = function(){//WIP: new matrix points to same storage as parent matrix
		var t = new Matrix(this.dimY,this.dimX);
		for(var i = 0; i<this.dimX; i++){
			for(var j = 0; j<this.dimY; j++){
				t.altElement(i,j,this.getElement(i,j));
			}
		}
		return t;
	}

	this.addColumns = function(n){//int->void
		for(var i=0;i<this.dimY;i++){
			for(var j=0;j<n;j++){
				this.val[i][this.dimX+j]=0;
			}
		}
		this.dimX += n;
	}

	this.addRows = function(n){//int->void
		for(var i=0;i<n;i++){
			this.val[this.dimY+i] = new Array();
			for(var j=0;j<this.dimX;j++){
				this.val[this.dimY+i][j] = 0;
			}
		}
		this.dimY+=n;
	}

	this.altRow = function(y,arr){//int,array->void
		if(typeof(arr)!="object" || arr.length!=this.dimX){
			console.warn("Warning: faulty array recieved, expected array of length " + this.dimX);
		}else if(y>this.dimY || y<0){
			console.warn("Warning: row does not exist");
		}else{
			this.val[y] = arr;
		}
	}

	this.altColumn = function(x,arr){//int,array->void
		if(typeof(arr)!="object" || arr.length!=this.dimY){
			console.warn("Warning: faulty array recieved, expected array of length " + this.dimY);
		}else if(x>this.dimX || x<0){
			console.warn("Warning: column does not exist");
		}else{
			for(var i=0;i<this.dimY;i++){
				this.val[i][x] = arr[i];
			}
		}
	}

	this.altElement = function(y,x,n){//int,int,number->void
		if(this.beautify){
			n = n;
		}
		this.val[y][x] = n;
	}

	this.dropRow = function(y){//int->void
		this.val.splice(y,1);
		this.dimY--;
	}

	this.dropColumn = function(x){//int->void
		for(row of this.val){
			row.splice(x,1);
		}
		this.dimX--;
	}

	this.getRow = function(y){//int->array
		return this.val[y];
	}

	this.getColumn = function(x){//int->array
		var col = new Array();
		for(var i=0;i<this.dimY;i++){
			col.push(this.val[i][x]);
		}
		return col;
	}

	this.getElement = function(y,x){//int,int->number
		return this.val[y][x];
	}

	this.mAdd = function(m){//Matrix->Matrix
		if(typeof(m)!="object"){
			console.warn("Warning: the matrix you want to add is not a matrix");
		}else if(!m.isMatrix(this.dimX,this.dimY)){
			console.warn("Warning: cannot add matrices of diffrent sizes");
		}else{
			var sum = new Matrix(this.dimX,this.dimY);
			for(var i=0;i<this.dimX;i++){
				for(var j=0;j<this.dimY;j++){
					sum.altElement(j,i,(this.getElement(j,i)+m.getElement(j,i)));
				}
			}
			return sum;
		}
	}

	this.cMultiply = function(c){//number->Matrix
		var cm = new Matrix(this.dimX,this.dimY);
		for(var i=0;i<this.dimY;i++){
			for(var j=0;j<this.dimX;j++){
				cm.altElement(i,j,(this.getElement(i,j)*c));
			}
		}
		return cm;
	}

	this.leftMultiply = function(m){//Matrix->Matrix//The owner matrix is on the right, the guest matrix on the left.
		if(m.dimX!=this.dimY || typeof(m)!="object"){
			console.warn("Warning: These matrices cannot be multiplied");
		}else{
			var mn = new Matrix(m.dimY,this.dimX);
			for(var n=0;n<m.dimY;n++){
				for(var p=0;p<this.dimX;p++){
					var j=0;
					for(var i=0;i<this.dimY;i++){
						j+=(m.getElement(n,i)*this.getElement(i,p));
					}

					mn.altElement(n,p,j);
				}
			}

		}
		return mn;
	}

	this.rightMultiply = function(m){//Matrix->Matrix//The owner matrix is on the left, the guest matrix on the right.
		return m.leftMultiply(this);
	}

	this.transpose = function(){//->Matrix
		var t = new Matrix(this.dimX,this.dimY)
		for(var i=0;i<this.dimY;i++){
			for(var j=0;j<this.dimX;j++){
				t.altElement(j,i,this.getElement(i,j));
			}
		}
		return t;
	}

	this.trace = function(){//->number
		var tr = 0;
		var c = Math.min(this.dimX,this.dimY);
		for(var i=0;i<c;i++){
			tr+=this.getElement(i,i);
		}
		return tr;
	}

	this.minor = function(i){//int->Matrix
		var c = this.clone();
		c.dropRow(0);
		c.dropColumn(i);
		return c;
	}

	this.trueminor = function(posy,posx){//int,int->Matrix
		var c = this.clone();
		c.dropRow(posy);
		c.dropColumn(posx);
		return c;
	}

	this.det = function(){//->number
		if(this.dimY!=this.dimX){
			console.error("Cannot give a determinant of a non-square matrix");
			return false;
		}else if(this.dimY == 2){
			return ((this.getElement(0,0)*this.getElement(1,1))-(this.getElement(1,0)*this.getElement(0,1)));
		}else{
			var d = 0;
			var minor = undefined;
			for(var i=0;i<this.dimY;i++){
				minor = this.minor(i); 
				
				if(i%2==0){
					d+=(minor.det()*this.getElement(0,i));
					
				}else{
					d-=(minor.det()*this.getElement(0,i));
					
				}
			}
			return d;

		}
	}

	this.inverse = function(){//->Matrix
		if(this.dimX!=this.dimY){
			console.error("Cannot return an inverse matrix of a non-square matrix");
			return false;
		}else if(this.det()==0){
			console.warn("Cannot return inverse, matrix is singular");
			return false;
		}else{
			var inv = new Matrix(this.dimX,this.dimY);
			for(var i=0;i<this.dimY;i++){
				for(var j=0;j<this.dimX;j++){
					
					var c = this.trueminor(j,i);
					
					if((j+i)%2==0){
						c = c.det();
					}else{
						c = c.det()*-1;
					}
						;
					c /= this.det();
					
					inv.altElement(i,j,c);
				}
			}
		}
		return inv;//this.leftMultiply(inv).val -> [I]
	}



	for(var i=0;i<y;i++){//init
		this.val[i] = new Array();
		for(var j=0;j<x;j++){
			this.val[i][j] = 0;
		}
	}
}



/*TODO
diagonalizations?
*/


var n = new Matrix(2,2);//debug dummies
n.altRow(0,[1,2]);
n.altRow(1,[3,4]);

var m = new Matrix(3,3);
m.altRow(0,[1,2,3]);
m.altRow(1,[4,8,6]);
m.altRow(2,[7,-2,9]);

var q = new Matrix(4,4);
q.altRow(0,[2,-1,4,5]);
q.altRow(1,[-1,0,0,3]);
q.altRow(2,[4,3,5,-3]);
q.altRow(3,[2,2,-3,0]);

//open question: where does "this" refer to?

File: Java/Matrix/src/Matrix.java
import java.util.Arrays;

public class Matrix {
    public int x;
    public int y;
    public double[][] val;

    public Matrix(int x, int y){
        this.x = x;
        this.y = y;
        this.val = new double[y][x];
    }

    @Override
    public String toString() {
        StringBuilder string = new StringBuilder("Matrix:{\n");
        for(double[] row : this.val){
            string.append(Arrays.toString(row)).append("\n");
        }
        string.append("}");
        return string.toString();
    }


    public static Matrix clone(Matrix m){
        Matrix out = new Matrix(m.x,m.y);
        out.construct(m.val);
        return out;
    }


    public boolean construct(double[][] value){
        if(value.length != this.y){
            return false;
        }else{
            for(double[] row : value){
                if(row.length != this.x){
                    return false;
                }
            }
        }
        this.val = value;
        return true;
    }

    public void addColumns(int n){
        double[][] temp = this.val;
        this.x+=n;
        this.val = new double[this.y][this.x];
        for(int i = 0; i<temp.length; i++){
            for(int j = 0; j<temp[i].length; j++){
                this.val[i][j] = temp[i][j];
            }
        }
    }

    public void addRows(int n){
        double[][] temp = this.val;
        this.y+=n;
        this.val = new double[this.y][this.x];
        for(int i = 0; i<temp.length;i++){
            for(int j = 0; j<temp.length;j++){
                this.val[i][j] = temp[i][j];
            }
        }
    }

    public void setRow(int y, double[] arr){
        this.val[y] = arr;
    }

    public void setColumn(int x, double[] arr){
        for(int i = 0; i<arr.length; i++){
            this.val[i][x] = arr[i];
        }
    }

    public void setElement(int x, int y, double n){
        this.val[y][x] = n;
    }

    public void dropColumn(int x){
        double[][] temp = this.val;
        this.x--;
        this.val = new double[this.y][this.x];
        for(int i = 0; i<temp.length; i++){
            for(int j = 0; j<temp[i].length; j++){
                if(j==x){
                    continue;
                }else if(j<x){
                    this.val[i][j] = temp[i][j];
                }else{
                    this.val[i][j-1] = temp[i][j];
                }
            }
        }
    }

    public void dropRow(int y){
        double[][] temp = this.val;
        this.y--;
        this.val = new double[this.y][this.x];
        for(int i = 0; i<temp.length; i++){
            for(int j = 0; j<temp[i].length; j++){
                if(i==y){
                    continue;
                }else if(i<y){
                    this.val[i][j] = temp[i][j];
                }else{
                    this.val[i-1][j] = temp[i][j];
                }
            }
        }
    }

    public double[] getRow(int y){
        return this.val[y];
    }

    public double[] getColumn(int x){
        double[] col = new double[this.y];
        for(int i = 0; i<this.val.length; i++){
            col[i] = val[i][x];
        }
        return col;
    }

    public double getElement(int x, int y){
        return this.val[y][x];
    }

    public static Matrix add(Matrix m1, Matrix m2) throws Exception {
        if(m1.x != m2.x || m1.y != m2.y){
            throw new Exception("Cannot add two matrices of different sizes");
        }else{
            Matrix sum = new Matrix(m1.x,m2.y);
            for(int i = 0; i<m1.y; i++){
                for(int j = 0; j<m1.x; j++){
                    sum.setElement(j,i,(m1.getElement(i,j)+m2.getElement(i,j)));
                }
            }
            return sum;
        }
    }

    public static Matrix multiply(int c, Matrix m){
        Matrix prod = new Matrix(m.x,m.y);
        for(int i = 0; i<m.y; i++){
            for(int j = 0; j<m.x; j++){
                prod.setElement(j,i,(m.getElement(i,j)*c));
            }
        }
        return prod;
    }

    public static Matrix multiply(Matrix left, Matrix right) throws Exception{
        if(left.x != right.y) {
            throw new Exception("Unable to multiply matrices: the amount of columns on the left has to be the same as the amount of rows on the right.");
        }else{
            Matrix prod = new Matrix(right.x,left.y);
            for(int i = 0; i<left.y; i++){
                for(int j = 0; j<right.x; j++){
                    double s = 0;
                    for(int k = 0; k<right.y;k++){
                        s+=(left.getElement(k,i)*right.getElement(j,k));
                    }
                    prod.setElement(j,i,s);
                }
            }
            return prod;
        }
    }

    public static Matrix transpose(Matrix m){
        Matrix t = new Matrix(m.y,m.x);
        for(int i = 0; i<m.y; i++){
            for(int j = 0; j<m.x; j++){
                t.setElement(i,j,m.getElement(j,i));
            }
        }
        return t;
    }

    public static double trace(Matrix m){
        double tr = 0;
        int l = Math.min(m.x,m.y);
        for(int i = 0; i<l; i++){
            tr+=m.getElement(i,i);
        }
        return tr;
    }

    public static Matrix minor(int x, int y, Matrix m){
        Matrix minor = Matrix.clone(m);
        minor.dropRow(y);
        minor.dropColumn(x);
        return minor;
    }

    public static double det(Matrix m) throws Exception {
        if(m.x != m.y){
            throw new Exception("Cannot give a determinant of a non-square matrix");
        }else if(m.x == 2){
            return ((m.getElement(0,0)*m.getElement(1,1))-(m.getElement(1,0)*m.getElement(0,1)));
        }else{
            double d = 0;
            Matrix minor;
            for(int i = 0; i<m.y; i++){
                minor = Matrix.minor(i,0,m);
                if(i%2==0){
                    d+=(Matrix.det(minor)*m.getElement(i,0));
                }else{
                    d-=(Matrix.det(minor)*m.getElement(i,0));
                }
            }
            return d;
        }
    }

    public static Matrix inverse(Matrix m) throws Exception {
        if(m.x!=m.y){
            throw new Exception("Cannot take an inverse of a non-square matrix");
        }else if(Matrix.det(m)==0){
            throw new Exception("Cannot take inverse: matrix is singular");
        }else if(m.x==2){
            Matrix inv = new Matrix(2,2);
            inv.setElement(0,0,(m.getElement(1,1)/Matrix.det(m)));
            inv.setElement(1,0,(m.getElement(1,0)/Matrix.det(m)*-1));
            inv.setElement(0,1,(m.getElement(0,1)/Matrix.det(m)*-1));
            inv.setElement(1,1,(m.getElement(0,0)/Matrix.det(m)));
            return inv;
        }else {
            Matrix inv = new Matrix(m.x, m.y);
            for (int i = 0; i < m.y; i++) {
                for (int j = 0; j < m.x; j++) {
                    Matrix c = Matrix.minor(i, j, m);
                    if ((i + j) % 2 == 0) {
                        inv.setElement(j, i, Matrix.det(c) / Matrix.det(m));
                    } else {
                        inv.setElement(j, i, Matrix.det(c) / Matrix.det(m) * -1);
                    }
                }
            }
            return inv;
        }
    }

    static class Identity extends Matrix{
        public Identity(int n) {
            super(n,n);
            for(int i = 0; i<n; i++){
                this.setElement(i,i,1);
            }
        }
    }

}
